setup.json

```diff
@@ -19,105 +19,87 @@
-    "paramiko==2.6.0",
-    "passlib==1.7.1",
-    "pika==1.1.0",
-    "plumpy==0.14.3",
-    "psutil==5.6.3",
-    "psycopg2-binary==2.8.3",
-    "pyblake2==1.1.2; python_version<'3.6'",
-    "python-dateutil==2.8.0",
-    "pytz==2019.3",
-    "pyyaml==3.13",
-    "reentry>=1.3.0",
-    "simplejson==3.16.0",
-    "sqlalchemy-utils==0.34.2",
-    "sqlalchemy==1.3.10",
-    "tabulate==0.8.5",
+    "jinja2~=2.10",
+    "kiwipy[rmq]~=0.5.1",
+    "numpy~=1.17",
+    "paramiko~=2.6",
+    "pika~=1.1",
+    "plumpy~=0.14.3",
+    "psutil~=5.6",
+    "psycopg2-binary~=2.8",
+    "pyblake2~=1.1; python_version<'3.6'",
+    "python-dateutil~=2.8",
+    "pytz~=2019.3",
+    "pyyaml~=3.13",
+    "reentry~=1.3",
+    "simplejson~=3.16",
+    "sqlalchemy-utils~=0.34.2",
+    "sqlalchemy~=1.3",
+    "tabulate~=0.8.5",
     "tornado<5.0",
-    "tzlocal==2.0.0",
-    "upf_to_json==0.9.2",
-    "uritools==2.2.0",
-    "wrapt==1.11.2"
+    "tzlocal~=2.0",
+    "upf_to_json~=0.9.2",
+    "wrapt~=1.11"
   ],
   "extras_require": {
     "ssh_kerberos": [
-      "gssapi==1.6.1",
-      "pyasn1==0.4.7"
+      "gssapi~=1.6",
+      "pyasn1~=0.4.8"
     ],
     "rest": [
-      "flask-cache==0.13.1",
-      "flask-cors==3.0.8",
-      "flask-httpauth==3.3.0",
-      "flask-marshmallow==0.10.1",
-      "flask-restful==0.3.7",
-      "flask-sqlalchemy==2.4.1",
-      "flask==1.1.1",
-      "itsdangerous==1.1.0",
-      "marshmallow-sqlalchemy==0.19.0",
-      "pyparsing==2.4.2",
-      "python-memcached==1.59",
-      "seekpath==1.9.3",
-      "sqlalchemy-migrate==0.12.0"
+      "flask-cors~=3.0",
+      "flask-restful~=0.3.7",
+      "flask~=1.1",
+      "pyparsing~=2.4",
+      "python-memcached~=1.59",
+      "seekpath~=1.9"
     ],
     "docs": [
       "docutils==0.15.2",
-      "jinja2==2.10.3",
-      "markupsafe==1.1.1",
-      "pygments==2.4.2",
-      "sphinx-rtd-theme==0.4.3",
-      "sphinx==2.2.0",
-      "sphinxcontrib-contentui==0.2.4",
-      "sphinxcontrib-details-directive==0.1.0"
+      "pygments~=2.5",
+      "sphinx-rtd-theme~=0.4.3",
+      "sphinx~=2.2",
+      "sphinxcontrib-contentui~=0.2.4",
+      "sphinxcontrib-details-directive~=0.1.0"
     ],
     "atomic_tools": [
-      "PyCifRW==4.4.1",
-      "ase==3.17.0",
-      "monty==2.0.4",
-      "pymatgen<=2018.12.12",
-      "pymysql==0.9.3",
-      "seekpath==1.9.3",
-      "spglib==1.14.1.post0"
+      "PyCifRW~=4.4",
+      "ase~=3.18",
+      "pymatgen>=2019.7.2",
+      "pymysql~=0.9.3",
+      "seekpath~=1.9",
+      "spglib~=1.14"
     ],
     "notebook": [
       "jupyter==1.0.0",
       "notebook<6"
     ],
     "testing": [
       "aiida-export-migration-tests==0.8.0",
-      "codecov==2.0.15",
-      "coverage==4.5.4",
-      "pg8000<1.13.0",
-      "pgtest==1.3.1",
-      "pytest==4.6.6",
-      "pytest-cov==2.8.1",
-      "sqlalchemy-diff==0.1.3"
+      "pg8000~=1.13",
+      "pgtest~=1.3",
+      "pytest~=5.3",
+      "sqlalchemy-diff~=0.1.3"
     ],
     "dev_precommit": [
```

Comment:
I think for the pre-commit dependencies it actually makes sense to pin them exactly. The issue with these is that even a "backwards compatible" change in behavior can cause the build to fail. 

For example, when ``pylint`` has a false positive, we add a ``# pylint: disable=...`` comment. If a subsequent patch fixes the false positive, it will trigger ``useless-suppression`` errors.

Reply:
Also, they are presumably used when developing ``aiida-core`` - so there really isn't much of a concern for being compatible with other environments / packages.

Reply:
We have the `useless-suppression` warning disabled globally, but I see your point. This can also just happen with `pre-commit` and `yapf` that introduce new formatting rules. However, if that happens, since these are only for developers anyway, the dev just upgrades his dependencies in his development environment and commits the changes no?

Reply:
Just imagine your PR fails because in just that moment some new yapf/pylint version has been released...
I would prefer to avoid this ;-)

Reply:
Exactly, it can be quite an annoying thing. Also, if we have scheduled CI on older branches they will just suddenly fail with pre-commit errors, for no reason.